docs/source/configuration/validation.md

```diff
@@ -0,0 +1,204 @@
+  name: 'fieldValidator',
+  dependencies: ['phoneNumber', 'isValidPhone'],
+  component: phoneValidator,
+});
+```
+
+It takes two `dependencies` since we can potentially have several validators for the same `widget`.
+The first element should be the name of the `widget`, and the second you can set it up to identify the validator.
+You should specify the `widget` in the JSON schema of the block (or as additional data in the content type definition).
+The next example is for the use case of a block JSON schema:
+
+```ts
+let blockSchema = {
+  // ... fieldset definition in here
+  properties: {
+    ...schema.properties,
+    phone: {
+      title: 'Phone number',
+      description: '',
+      widget: 'phoneNumber',
+    },
+  },
+  required: [],
+};
+```
+
+### Per behavior and field name validator
+
+These validators are applied depending on the behavior (usually coming from a content type definition) in combination with the name of the field.
+
+```ts
+config.registerComponent({
+  name: 'fieldValidator',
+  dependencies: ['plone.eventbasic', 'start', 'startValidator'],
+  component: startEventDateRangeValidator,
+});
+```
+
+The first dependency should be the name of the behavior, and the second the name (`id`) of the field.
+It can get a third dependency in case you want to specify several validators for the same behavior - field id combination.
+This type of validator only applies to content-type validators.
+
+### Per block type and field name validator
+
+These validators are applied depending on the block type in combination with the name of the field in the block settings JSON schema.
+
+```ts
+config.registerComponent({
+  name: 'fieldValidator',
+  dependencies: ['slider', 'url', 'isURL'],
+  component: urlValidator,
+});
+```
+
+The first dependency should be the `id` of the block, and the second the `id` of the field.
+It can get a third dependency in case you want to specify several validators for the same block type - field id combination.
+This type of validator only applies to blocks.
+
+### Specific validator using the `validator` key in the field
+
+A final type of validator is applied to the field if the `validator` key is present in the JSON schema definition of the form field.
+
+```ts
+config.registerComponent({
+  name: 'fieldValidator',
+  dependencies: ['isURL'],
+  component: urlValidator,
+});
+```
+
+The dependencies take one single name, in this case, the name of the validator.
+You should specify the validator in the JSON schema of the block (or as additional data in the content type definition).
+
+```ts
+let blockSchema = {
+  // ... fieldset definition in here
+  properties: {
+    ...schema.properties,
+    customField: {
+      title: 'Default field',
+      description: '',
+      validator: 'isURL',
+    },
+  },
+  required: [],
+};
+```
+
+It does not need to be tied to any field `type` or `widget` definition.
+It runs in addition to all the above, so it complements the other validators if any apply.
```

Comment:
My biggest piece of feedback on this overall is using an array for the `dependencies` feels like trying to fit a square peg into a round hole. It's quite hard to keep track of all the different ways validators can be registered and which values need to go in the dependencies for each one. And it also seems to be like there might be some ambiguous cases, like if there is a field type and a widget type with the same name.

I don't know if the registry supports this but would it be possible to make things a bit more explicit by specifying the dependencies as an object instead of an array? So for example:
```
config.registerComponent({
  name: 'fieldValidator',
  dependencies: {
    blockType: 'slider',
    fieldName: 'url',
    validatorName: 'isURL',
  },
  component: urlValidator,
});
```

Reply:
@davisagli TBH, I'm not fully satisfied with it... I thought that the reason was mainly because we can't get it to work as satisfying as the ZCA works :) I envisioned the resolution algorithm using a string and concating the deps if we have an array, so I've played with that concept, even in the new `getComponents` registry API method.

Using an object would be possible too, I guess. I can play with the idea, and leave the original API in place also.

Reply:
@davisagli I did an initial implementation of using an object instead of an array but I have doubts on how to hash the resultant key to store that object, so then `getComponents` work in the same simple way, under the same concepts that we had before. It does work, but under some assumptions that I do not like or they are not satisfactory/safe :/ 

Maybe if you have some time, we can talk about it.